Tasks/Npm/npmtask.ts

```diff
@@ -4,89 +4,113 @@ import path = require('path');
 import url = require('url');
 import tl = require('vsts-task-lib/task');
 import trm = require('vsts-task-lib/toolrunner');
+import nutil = require('nuget-task-common/utility');
+
 var extend = require('util')._extend;
 
 interface EnvironmentDictionary { [key: string]: string; }
 
-tl.setResourcePath(path.join( __dirname, 'task.json'));
+tl.setResourcePath(path.join(__dirname, 'task.json'));
 
 executeTask();
 
 async function executeTask() {
 
-    var cwd = tl.getPathInput('cwd', true, false);
-    tl.mkdirP(cwd);
-    tl.cd(cwd);
-
-    var command = tl.getInput('command', true);
-    if (command.indexOf(' ') >= 0) {
-        tl.setResult(tl.TaskResult.Failed, tl.loc('InvalidCommand'));
-        return;
+    try {
+        var filePath = tl.getPathInput('cwd', true, false);
+        var dirList: string[] = [];
+        if (filePath.indexOf("*") !== -1 || filePath.indexOf("?") !== -1) {
+            var filesList = nutil.resolveFilterSpec(filePath, tl.getVariable("Build.SourceDirectory"));
+            tl.debug("number of files matching filePath: " + filesList.length);
+            for (var file of filesList) {
+                dirList.push(path.dirname(file));
+            }
+        }
+        else if (fs.statSync(filePath).isFile()) {
+            dirList.push(path.dirname(filePath));
+        }
+        else {
+            dirList.push(filePath);
+        }
     }
-
-    var npmRunner = tl.tool(tl.which('npm', true));
-    npmRunner.arg(command);
-    npmRunner.line(tl.getInput('arguments', false));
-
-
-    if(shouldUseDeprecatedTask()) {
-
-        // deprecated version of task, which just runs the npm command with NO auth support.
-        try{
-            var code : number = await npmRunner.exec();
-            tl.setResult(code, tl.loc('NpmReturnCode', code));
-        } catch (err) {
-            tl.debug('taskRunner fail');
-            tl.setResult(tl.TaskResult.Failed, tl.loc('NpmFailed', err.message));
+    catch (error) {
+        tl.warning(error);
+    }
+    for (var cwd of dirList) {
+        tl.debug('Executing command in directory: ' + cwd);
+        tl.mkdirP(cwd);
+        tl.cd(cwd);
+
+        var command = tl.getInput('command', true);
+        if (command.indexOf(' ') >= 0) {
+            tl.setResult(tl.TaskResult.Failed, tl.loc('InvalidCommand'));
+            return;
```

Comment:
Pull all the code which does not change across the iterations, out of this for loop. For example setting up the npm command.